pkg/auth/handler/reset_password.go

```diff
@@ -0,0 +1,148 @@
+}
+
+func (f ResetPasswordHandlerFactory) ProvideAuthzPolicy() authz.Policy {
+	return policy.AllOf(
+		authz.PolicyFunc(policy.DenyNoAccessKey),
+		authz.PolicyFunc(policy.RequireAuthenticated),
+		authz.PolicyFunc(policy.DenyDisabledUser),
+	)
+}
+
+type ResetPasswordRequestPayload struct {
+	AuthInfoID string `json:"auth_id"`
+	Password   string `json:"password"`
+}
+
+func (p ResetPasswordRequestPayload) Validate() error {
+	if p.AuthInfoID == "" {
+		return skyerr.NewInvalidArgument("invalid auth id", []string{"auth_id"})
+	}
+
+	if p.Password == "" {
+		return skyerr.NewInvalidArgument("empty password", []string{"password"})
+	}
+
+	return nil
+}
+
+// ResetPasswordHandler handles signup request
+type ResetPasswordHandler struct {
+	PasswordChecker      dependency.PasswordChecker `dependency:"PasswordChecker"`
+	TokenStore           authtoken.Store            `dependency:"TokenStore"`
+	AuthInfoStore        authinfo.Store             `dependency:"AuthInfoStore"`
+	PasswordAuthProvider password.Provider          `dependency:"PasswordAuthProvider"`
+}
+
+func (h ResetPasswordHandler) DecodeRequest(request *http.Request) (handler.RequestPayload, error) {
+	payload := ResetPasswordRequestPayload{}
+	err := json.NewDecoder(request.Body).Decode(&payload)
+	return payload, err
+}
+
+func (h ResetPasswordHandler) Handle(req interface{}, _ context.AuthContext) (resp interface{}, err error) {
+	payload := req.(ResetPasswordRequestPayload)
+
+	authinfo := authinfo.AuthInfo{}
+	if e := h.AuthInfoStore.GetAuth(payload.AuthInfoID, &authinfo); e != nil {
+		if err == skydb.ErrUserNotFound {
+			// logger.Info("Auth info not found when setting disabled user status")
+			err = skyerr.NewError(skyerr.ResourceNotFound, "User not found")
+			return
+		}
+		// logger.WithError(err).Error("Unable to get auth info when setting disabled user status")
+		err = skyerr.NewError(skyerr.ResourceNotFound, "User not found")
+		return
+	}
+
+	if err = h.PasswordChecker.ValidatePassword(audit.ValidatePasswordPayload{
+		PlainPassword: payload.Password,
+	}); err != nil {
+		return
+	}
+
+	// reset password
+	principal := password.Principal{}
+	err = h.PasswordAuthProvider.GetPrincipalByUserID(authinfo.ID, &principal)
+	if err != nil {
+		if err == skydb.ErrUserNotFound {
+			err = skyerr.NewError(skyerr.ResourceNotFound, "user not found")
+			return
+		}
+		return
+	}
+	principal.PlainPassword = payload.Password
+	err = h.PasswordAuthProvider.UpdatePrincipal(principal)
+	if err != nil {
+		return
+	}
+
+	// generate access-token
+	token, err := h.TokenStore.NewToken(authinfo.ID)
+	if err != nil {
+		panic(err)
+	}
+
+	if err = h.TokenStore.Put(&token); err != nil {
+		panic(err)
+	}
+
+	authContext := context.AuthContext{}
+	authContext.AuthInfo = &authinfo
+
+	resp = response.NewAuthResponse(authContext, skydb.Record{}, token.AccessToken)
+
+	// Populate the activity time to user
+	now := timeNow()
+	authContext.AuthInfo.LastLoginAt = &now
```

Comment:
According to the old skygear code, `LastLoginAt` and `LastSeenAt` does not have to be set.